{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c7dac45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of reads: 4133\n",
      "Total number of reads after processing: 2678\n",
      "Before Remedy: \n",
      "Total number of valid reads: 2678\n",
      "Exact matched sequences before RS correction: 2573\n",
      "Matched sequences after RS correction: 2649\n",
      "['CGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACCGACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTAGTAAGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACCTACGTACGTACGTACGTACGTACGTACGACT', 'ACGTACGTACGTACTTCCGTCCGTACGTAGGTACGTACGTACGTACGAACGTACGTAGGTAGGTACGTACGTACGTCCGTACGAACTTACGTACGTACGTACGTAGGTCCGTCCGAACGTACGTACGTCCTTACGTAGGTACGTACGAACGTACTTACGTAGGTACGTACGTACGAACGTACTTACGTACGAACGTACGTACGTACGTACGTACGTACTTACGTACGTACGTAGTTACGTACGTACGTACGTACGTAGTAACGTACGTACTTACGTACGAACTAACGTACGTCCGTACGTACGTACGTACGTACGTCCGTACGTACGTACGTACGACCGTAGGTACGAACGTCCGAACTTACGTACGTACGTACTTCCGTACGTACGAACGTACGTACGTCGGTACGTACGTCCTAACGAAGTAACGTACGTACTTACGTACGTACGTAGGAAGGTAGGAACTTCGGTACGTAGGAACGTACGACGGTCGTTACGTACGTACGTACGTACGTACTTACGTAGGTAGGTACGTACGTAGTTACGTACGAACTTCGGTACGAACGTCGTAACGTACTTACGTACGTCGGACCTTACGTAGTTACGTAGGTACGTCGGAAGGTAGGTACTTCCGTACGAACGTCGGTAGTACGTACGTACGTACGTACGTACGTAACTTACTACGTTCGGTGCGTCGTACGGTACTTCGTTCGGACGGACACT', 'CCGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTCCGTACGTACGTACGTACGTACGTACGTACGAACGTACGTACGTACGTACGTACGTACGTACTTACGTAGGTACGTACGTCCGTACGTACGTACGTACGTACGTACGTACGTACGTACGAACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGAACGTACGTACGTACGTCCGTACGTACGAAGGTCCGTACGTACGTACGTACGTACGTACGTACTTACGTACGAACGTACGTACTTACGTACGTACGTACGTCCGTACGTACGTCCGTACGAACTTACGTAGGTCCTTACGTCGGTACTAAGGTACGTACTTACGTACTTCGGTCCGTACGAAGGTCCGAACTTACGAAGGTCCGTACGTCCGTACGTACTAACGTACGTCCTTCCGTACGTACGTAGTTCCTACCGACGGACGGTAGGTACTACCGTACGTCGTAACGAACGACGTAACGTCCGAACTTCCTTCGGTACGTCCGACGGTACTAAGGAACTTACGTCCGTCCGTAGTTACTTCCGAACGTCCTTCCTTACGTACGTACTTACTTACTTCCGTACGTACGTACGTACGTACGTACGTCGGACGGTACGAACTAAGTAACGTACTACGGAACGAACTTACTTAACT', 'CGGAAGGTCCTTCCTACCGACGTTAGTACGTAAGGACCTACGGAAGGACGTACGTACGTTCGTAACTACGTAAGTACGTAAGTACCTACGTTAGGACGGACGTTCGGTCGTTCGTACGTAAGTACCTTCCTACCTACGTAAGTTCGTACCTTCGGACGGTAGTTACTACGGTCGGACGGACGTACCTACGGAACTAACTACGGACGTTACTTCGTTAGTTCGTAACTACCTAAGTACCTTAGTAAGTACGGAACGTCGTACGTTCCGAAGTACGTAAGTAAGGAAGGAACTACGTTAGGACGTACGGTCGTTACTACGTTCGGACGGAAGGACCTACGGACGGACCTTCCTTAGTAACTTCCTACGTAACGAAGTACGGTAGGTCCGAAGTACGGTCCTAAGTTAGTTAGTTCGTTCCGAACGAACTACGTAAGTACGTAAGTAAGGAAGGAAGTACGGACCGTCGTAACTTCGGACGTTCGTTCGGTCGGACCGTCGGACCTACGTTCCGACCTACGGTCGGACCGTACTTAGTACGTTCGTACGGTCGTAACTTACGACGGTCCGTCCGACGGAAGTTACGACGTTCGTAACTACGTTAGTAACTTCGTAACTTAGTACGTTAGTTCCTTCCTAAGTTGGATCGACGCGACGGCCTAAGTAAGTGTGGACCCTCCGTCCTATGGAAGGAAGCGTACGTGCGAACGTGTTTCTTATACT', 'CGTTCCTAAGTAAGTTCGTACGTACCGAAGTACGTTCGTTAGTTCGTACGTTCGTTAGGACGTTCGGACGGTCCGTCCGTCCTAAGTTCGTACGTTCCGACCTAAGTACCTAAGGAAGGTCCTTAGTACGGTCCGACCTTCGGACCGACGGAACTTCGTACGTAAGTACGGAACGACGTACGTACGTACGTACGTACCTTAGTAAGTACGTAACTACGTACCTACGTTCGGTCCGTCGGTCCTACCTAACTACGTAACTTCCGAAGTACGTTACGACGGTCGTAACGACCTTAGGACCGACCGTACGACCGTCGTTCCTACGTTAGTTCCTTACTTACGTCCTAACTACGTACGTTCGTAAGGTCCGACGTACCGTACTACCTAAGTTCGTAACGTAGGACCGAACTAAGGAAGTACGTTCGGACGTAAGTTCCGTACTAAGGACCTTCCGTCGTACCGTACTACCGTCGGTCGTTCGGACCTAACTTAGTACGTTCCTTACTACGGACGTTCGTTCGTACGTAACTTCGTACCGAACGAACTTCGTACCGTCGTACCTTAGTACGTACGTTCGTTAGGACGGTACGTACGACGGTACGACCTAACTAACGTCGGTCCGACCTACCTTCGTACGTACGTAGCACATTTGTTCGCTAGGACCGGTGCTCATTTAGATGCGCAGTAGCTTATATGTGAGGTCAGGTCTGCCTACAGATAACT', 'ACGTCCGTACTACGTTCCTTCGGAACTTAGTACCGAAGTAAGGAAGTACGTAAGGTAGGAAGTAAGTTCCGTACGTCGTAACGAAGTTCGTACGGAACTAACGTCGTACCGAAGGAAGGTACTTAGGTCGGTCGGTACGACCTACCGAACGTCCTACCTTCGTTCGGACCTTCCTACCTACCTTACTTACGACGTTAGGTCCTTACGTCGGTCCGTCCTAAGGTAGTAAGGTACGAAGGAACGACCGTAGGACCTAACGTACGACCTACCTTAGGTACTTCGGACCTTCGGTCCTTCCTACCTTCCTACGTAAGGTCGGACGTTCCGTACTAAGGTACTAACGAAGTACGGTCCGACGTAAGTTACTACCTTAGTAACTAACTTAGGTACGTACGTAGTTCGTACCTACCGACGGAAGGACGTACCTAACGACCGTCGGACGGTCGTTACTACGGAACTTCGTTCGTTCCTAAGTTCCGACCGACGGACGGTCCTAAGGACGGACCGACCGACCTAAGGAAGTTAGTACCTTAGGTACGAAGGTAGTAACGAACTTCGTTACTACGGTCGGTCGGTCCGAAGTTCCGAACGTCGGAACTTCCGACGGACGTTCGTACCGAACTTCCTAACGACGGTCCGTCCAAATTTCTGTGCTACGAAGCTCGTTTAGAAACATACAAAGATCCGTCCAACGTCCGTCGTTCGTTTCGGCGTATAACT', 'CCTACCTAAGTTCCTTCGGTCGTTAGTAAGTTAGTAACTAACTAAGGACGGAACTTCGGACGTTCGTAAGGACCGTCGTACGTAACTTCCGTCCTACCTTCGTACGGTCCTACGGAAGGTACTAACGACCTAAGTTCGGAAGTTAGTACCTACCGTAGGTAGGACCGTAGTACGTAACGAACTTCGGTAGGAACTACGTTCGGACGTTCGGTCCTAAGTTAGTAACGACGGTCGTACGGTCGTAACTTCGTTAGTTCGGAACTTCCGACCTAACGTCGTAAGTAACTACGGACCTACCTAAGGACGTTACTACGTACGGTCCGACGTTCCGAAGGACCTAAGGAAGGAACGACCTAACTTACTTCCTTAGGACGTTCCGTACGAACGACGGAAGGTCCTAAGTTAGTTCGGACCTTCCGACGTAACTACCTAACTTCGTTACGACGGTCGTAAGTTAGTAAGGACGTTAGTAACTAACGTAGTTCGGACGGACGTACGGACGGTAGTTACTTACTAACTTCCGACGGAAGGACGTTACTTCGGAAGGTCCGTAGGAAGGTCCTACGTTCGTAACGTCGTACCTACCGTAGGTAGGTAGTAAGTTCCTAACGTCCTACGTTCGGTAGGAACTACCGTAGGAGCATAGATCGGACTTCGTACCCAAGTAAAGAAATTCCTTCCGATCTAAGCAACCATGGGTCCAAGCGTGTTTCGGTCACT', 'CCTACCGAAGGTACTACGTTCCGACGGTAGGTAGGTCGGTCGTACCGAACGTACGACGGAAGGTCCGTACTTCGTTCCGTCGGAACTTAGGTCCTACGGTCGGTAGTACGGAACGACCGAAGGAACGTCGGAAGGAACTTCCTACCTAACGTACGTCGGTAGTAACTTACGTCGGACGTTACTAACGTCGTAACGTACGTACGTAGGTCGGACGTTAGTAACGTACTAAGTTCCGTAGTACCTTCGTTCCTACCTTCCGTCCGAACTTCGTAAGTACGTTACTACGGTCGGACCTACGGTCCGAAGTAACGAAGTTCCTTAGTTCGTTACTAACGTAGGAACGTACTTACTTACGACCTTACTTCCGACCTTACTTACTAAGGAAGGTCCTTACTAACGAAGTACGGTACGACGGAACGACCGAACGTACTTCCGTCGGTCGTACGTTACGAAGGAACGTCGGACGGACGTTAGTTCCGTCGGACCGTCGGAAGGTAGTTAGGAAGTTACTTCCTTACTTAGGTACGTCGGACCTAACGACCGTACGTACTTCCGTACGAACTACCGTAGGACCGACGGAACTTCGTAACGAACTTACGTACGAACGAAGGAAGTTACTACGGTCGTTACGAAGGTAGTTGTGAGGTACTTTGCTAGGTCCCATCGGACTTACTTTATGCAGACCCATCGGTCCGAACGCACGTCGTCCGGACGGTCACT', 'CGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACCTACCTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTAGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGACT', 'ACGAACGTACGTACGTACGTACTTACGTACGTAGGTACGTCCTTACGTACGTCCGTACGTACGAACGTAGGTACGTACGTACGTACGTACGAACGAACTTACGTACGTACGTACGTACGTACGTCCTTACGTACGAACGTAGGTACGTACGTACTTAGGTACGTACGTACGTACGTACGACCGTACGAACGTACGTACTTACTTACGTACGTAGGTCCGTACGTCCGAAGGTACGTACGTACGTAGGTAGGTCCGTACGTCCGTACGAACGTACGTCCGTACTTACTTACGTACGAACGTACGAACTTACGAACGAACGTACGAACGTACGTAGGTACGTACGAAGGAAGGTCCGTAGGTACGTCCGTACGTACGTACTTAGGAACTTCCGTAGTAACGACCGAAGGAACTTAGGTCCGTAGTTACTTAGGTACTTACGTACGTACGTACGTCCGTACGTACGTACGTCCGTACGTACGTACTAAGGTAGTTCCGTAGGAACGTAGGAACGTCCGACGGTACTACGGTCCGTACTTACGAACTAACTTACGACGGTAGGTACTTCGTAACGTAGTTACGTAGGTACGAACGAAGGAACTACCGACGGTACTACGGAACGTCGGAACGTCCTAAGGTCGGTCGTAAGTACCTAAGTAACGTAGGAAGTAACGTAGGAACTTCGTACGTTACTTAGTACGGAACTTCGTTACGACGGACACT', 'ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACTTAGGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGAACGTACGTACGTACGTCCGTACGTACGTACGTACGTCCTTACGTACGTACGTACGTACGTACGTCCGTACGTACGTCCGTACGTACGTACGTCGGTAGGACCGAACGTACGTCGGAACGTACGTACGTACGAACGTACGTACGTCGTTACGTACGTACGTACGTACGTACGTACGTACGTACGTCCGTACGTACGTACGTACGTACTTAGTAAGGTAGGTACGTCCGAAGGTACGTACGTACGAACGTACGTACGTACGTACGTACGTACGTACTTCGGTAGGTACGAACGTACGTCCGTACGTACGTACGTACGTACGTACGTCCGTAGGTCCGTCGTACCTTCCGTCGTACCTTCGGAACGTACGACGGACCGTACTTCGGAACGACGTTACTTAACT', 'CGTAAGTACCTAAGTAAGTACCTACCGAACTAAGGACGGACCTACGTACCTTCCTAAGTAACGACCGTAGGACGTACGGACGTACCTTCGTACGGACCTACGGACCTACGGTCCTACGGTCGTACGTTCCTACGTACGTACGTACGTACGGTCCGAAGTACCTTCGTTAGGTCGTACGGACCTTAGTACCTTAGTACGTTCCTACCTACCTACGGACGGTAGGTCCTAAGTTCGGACGTAAGTACGTACGGACGGTCGTACCTTCGTAACGTCGGACGGAAGGACGTACGTTACGACCTTCCTACCTTCCGACGGAAGGTACGAAGGACCTACGTACCTACCGAAGGTCCTAACTAAGGACGTACGGTCGTTCCTAACTACGGAACTACCGACGGTCCGAAGTACGTAAGTACGGACCTAACTACCGTAGGTACGTCCTTACGAAGTTCCTAACTTCCGACCTAACTACCGTCGGACGGTCGGTCGTAACGTCCGAAGTACGTACGTACGTTCGGACGGACGTACGTTCGGTCGTTCGTACGTACGGACGTAAGGTACTACGTAAGGACGGACGGAAGTACGTACGTACCGACGTACCGAAGTAAGTACGTACCGTCCTTCCTAAGGACGGAAGGTCGGAGGACCCGTACTCGCTAGCGTGTACGCACGTACGGAAATACGCGAACTTCGTTCTGTAGGCAGACGGTACCTTACTCCACT', 'ACTACGTTAGGTCCTACCTAAGTAAGGACGTACGTTACTTCGTACCGTCGGACGTAAGGTAGTAAGTACCTACGTACGTAACGACGGACGTACCGTCGTTACTTCGGAAGTACGTAACTACCTACGGAAGTAAGGAAGTACCTTAGGACGTAAGGTACGTACGAAGGAACTACGGACGTACCGACGTTCGTAACGACGGACGGTAGGTCGTTACTAAGTACGTACGTAAGTAACTTACGACGGAACTTCGTTACTACCTTAGGTACGTCGTACGGTACTACCTTCCGAACGACGGTAGTTAGTTAGTAAGTTCCTACGGACGTTCGTTCGTTCCGACCGTAGTTCCTACGGTCGTACCTAACTAACGACCTAAGTTCGGAACTTCGGTAGTTACTAACTTCCGTACGTCGGACGTAAGTTCCGACGTACCTACGTAAGGAAGTACCGTAGGAACTACGGTAGTACGGAAGTAAGGACGTAACGACCGACCTTACTAACGTCGTAACGTACTTACGTACTTCGTTACTAAGTACGGAAGTTCGGTACGACGGTCCGTCCGACGGAAGTTAGTTACTTCCGTCCTTCGTTCGTTCCTTCGGTAGTAAGTTAGGTAGTAAGGAAGGTCCTACCGTAGTAACTTCTACGGACCCAAATGTGTTAATAACCGACCGCACATCCGACCTTATGAACGAATTCCCTTCGATACATAGTAGTGTTACT', 'CGTACCTAAGGTCGTTCCTTCCTTACTTACTTCGTTACGTCCTAACGTAGTTCGTTCGGTCCTACCTAACGAAGTTACGACCGTCCGAACTACGTTAGTTCCGTAGGTCCTACGTTACGTACTTCGGAAGTACGTACCGAACGAACTAACTTACTACCTACGTAACTTCCTAACGTCCTACGGACCTTCGTACGTACCGAAGTTACTAACTAACTTAGTTCCGTCCTACGGAAGGTACTAAGGTACGTAGGTACTAACGTACGAACTTCCGACCTAACTTAGGAACGTCGGTCGTTACGACCGTCCTTAGGTCGGAAGTTAGGTACGTAGTTACGACGGACGTTACTACGGTACGACCTAACGTAGTTAGGTAGTTCGGACCTTCGGTACGAACGTAGGTACGTCCGTCGGTAGGAAGTAAGTTACTACCGACCTTACGACGTTACGAACTACGTTCCTAAGTAAGGAAGTTACTTACGTAGTAACGACCGAAGGACCTAACGAACGTACTAAGTTACTAAGTAACGAACGAAGGTCGGTCCGTAGTTCGGTACTAAGGACCGTCCTTCCGACGTAACTTAGTTACGTAGGTACTTACTTCCTACGGTAGGTACTTCCGTAGTTACTTACGTCCTAAGTACTTCGCGCCCAAGCGCGCTTAGGACGGCCTGTACGTCCGAGCATCGTAATTTAGTACGTACGACACTAAGATCGTTGACT', 'CGTAAGGAACGAAGTAACTACCTAAGGAACGAAGGTAGTTCGTACGGACCTTCCTTCCTACGGACGGTCGTACGGTCCTACGGTAGGTAGTACGTACCTTCCGTCGGACCGTAGTACGTACCGACCTAACGAAGTACGGTAGTACGTTCGTACGTACGGACGTTAGTACCGTCGGTCGGACGTTCGGTCCTTCGGTAGGACGTAAGTACGTAAGGTAGTAACTACGGACGTTCGTACCTACCGTACTTAGTTAGTAAGTTCGTAACGAAGGTAGTACCGACGTTCCTTCGTAACTTCCTAAGTAACTTCCTACGGACCTACCTACCGACGTTACGAACTACCGAAGGTCGTTCGGTAGGAAGGACGTAAGTTCGTTACTTCGTAAGTACCTTAGTTCGTTCCTACCGACCTTCGGACCTAAGTAAGTTAGGAACGTCGGTCGTACCTAAGTTCCGACCGTAGTTCCGAAGTAAGTTCCTTCCGTACGTAGTAAGTTCCTAAGGAAGGAACGACGTTCGGACCGAAGTTCGTTACTAACGTCCGTCCGAACGACCTACCGTCGTTCGTACGGACGTTCGTTAGGTCCTTAGGTCGGAAGGTAGGAAGGAAGGACCGAACGTAGTAAGTTAGGACCGTAGGTCGAAGCTTCCTAGGTAGCTCATACATGAGTATGCGAACATGGGACTAAGTTTGTTCATAAGTTCAGACAGGTACAACACT', 'ACGACGGTCGTTACGAACGTCGTTACTAACGTAGGAACTTCCTTAGGTCGTACGGTACGACGGACCTTCGGTACGACCTTCGTAAGGAACGAACGTACTTACTAAGGAAGTACGGTCCGAACTTCCGACGGTACGAACTTACGTCCGAACTACGGTACGTCGGTCCGTACGAAGGTACGACCTAACGAACGAAGGTCCTACGGAACTAAGGTACTTCCGTAGTTCCGAAGGTCCTTACGTCGGACGTTAGGTCGTACGTTAGGTCCGAACTTCCGTCCTACCGTAGTTACTAAGGAAGTTCCGAAGTACCTAAGGAAGGAAGTAACGACGGAAGTTAGGTCCGTCCGTACTTACTTCGGTCGTTCCTTCCGTAGGACGTTCGTTCGGTAGTTACGTCCGAACGTAGTTCGGAAGTTAGGACCGTCGGTACTTAGTACGTTAGTTACTACCGTAGTAAGTTAGGTACTTAGTAACGAACGTCCGTCGTTCCGTCCGACGGTCCTTACTTACTTACGACGTTCCTTCGTAACTACGGTACGTACGTAGTAACGTAGGAAGGTACGTACTAAGTTACTTCGGTCCGTACGTACTTAGTAACGTACTAAGTAACTTCCGAACGACCTAACTTAGGACCTTACTTGTACACGAATTCACGAACAAAGACCCTAATGTATGCCCTTCGTTGCGCCCGACGTCCCATGCTTCTATCCGAACGAGACT']\n",
      "Stored 'referenceStrands' (list)\n",
      "Reference #0: Yes!\n",
      "Reference #1: Yes!\n",
      "Reference #2: Yes!\n",
      "Reference #3: Yes!\n",
      "Reference #4: Yes!\n",
      "Reference #5: Yes!\n",
      "Reference #6: Yes!\n",
      "Reference #7: Yes!\n",
      "Reference #8: Yes!\n",
      "Reference #9: Yes!\n",
      "Reference #10: Yes!\n",
      "Reference #11: Yes!\n",
      "Reference #12: Yes!\n",
      "Reference #13: Yes!\n",
      "Reference #14: Yes!\n",
      "Reference #15: Yes!\n",
      "Remedy Process: \n",
      "Total number of valid reads: 1378\n",
      "Exact matched sequences before RS correction: 823\n",
      "Matched sequences after RS correction: 894\n",
      "After Remedy: \n",
      "Total number of valid reads: 4056\n",
      "Exact matched sequences before RS correction: 3396\n",
      "Matched sequences after RS correction: 3543\n"
     ]
    }
   ],
   "source": [
    "# This program aims to decode Reference Pool\n",
    "# Procesing workflow:\n",
    "# 1. Identify and exclude PCR primers. In Reference Sequences, we used universal primer sets.\n",
    "# 2. RS correcting the Reference Sequence and extract Reference Sequence.\n",
    "# 3. Get Reference Number (the first four bases after forward primer) and Segment Number (the following four bases after sequence of Reference Number )\n",
    "# 4. Put together Reference Strands into Reference Pool.\n",
    "\n",
    "from reedsolo import RSCodec\n",
    "# 2 error correction codes\n",
    "rsc = RSCodec(2)\n",
    "\n",
    "def converter(seq):\n",
    "    converter = {'A': '00', 'C': '01', 'G': '10', 'T': '11'} \n",
    "    bases = list(seq) \n",
    "    bases = [converter[base] for base in bases] \n",
    "    return ''.join(bases)\n",
    "\n",
    "def deconverter(seq):\n",
    "    deconverter = {'00': 'A', '01': 'C', '10': 'G', '11': 'T'} \n",
    "    doubleBits = [seq[i:i+2] for i in range(0, len(seq), 2)]\n",
    "    doubleBits = [deconverter[doubleBit] for doubleBit in doubleBits] \n",
    "    return ''.join(doubleBits)\n",
    "\n",
    "def oligoToBase3(seq):\n",
    "    oligoToBase3 = {'G': '0', 'T': '1', 'A': '2', 'C': '3', 'N': '3'}\n",
    "    oligoToBase3Converted = []\n",
    "    for i in range (0, len(seq)):\n",
    "        if (oligoToBase3[seq[i]] == '3'):\n",
    "            return -1\n",
    "        else:\n",
    "            oligoToBase3Converted.append(oligoToBase3[seq[i]])\n",
    "            if seq[i] == 'C':\n",
    "                oligoToBase3 = {'G': '0', 'T': '1', 'A': '2', 'C': '3', 'N': '3'}\n",
    "            elif seq[i] == 'G':\n",
    "                oligoToBase3 = {'T': '0', 'A': '1', 'C': '2', 'G': '3', 'N': '3'}\n",
    "            elif seq[i] == 'T':\n",
    "                oligoToBase3 = {'A': '0', 'C': '1', 'G': '2', 'T': '3', 'N': '3'}\n",
    "            elif seq[i] == 'A':\n",
    "                oligoToBase3 = {'C': '0', 'G': '1', 'T': '2', 'A': '3', 'N': '3'}\n",
    "\n",
    "\n",
    "    return ''.join(oligoToBase3Converted)\n",
    "\n",
    "\n",
    "def ternaryToDecimal(n):\n",
    "    decimal = 0\n",
    "    n = ''.join(reversed(n))\n",
    "    for i in range (0, len(n)):\n",
    "        decimal += (int(n[i]))*(pow(3, i))\n",
    "    return decimal\n",
    "\n",
    "\n",
    "def most_frequent(List):\n",
    "    counter = 0\n",
    "    string = List[0]\n",
    "     \n",
    "    for i in List:\n",
    "        curr_frequency = List.count(i)\n",
    "        if(curr_frequency> counter):\n",
    "            counter = curr_frequency\n",
    "            string = i\n",
    "            \n",
    "    return string\n",
    "\n",
    "\n",
    "%store -r array_reference\n",
    "# Using readlines() to get the content of every line of FASTQ file.\n",
    "file_R1 = open('sample5_S5_L001_R1_001.fastq', 'r')\n",
    "file_R2 = open('sample5_S5_L001_R2_001.fastq', 'r')\n",
    "\n",
    "Lines_R1 = file_R1.readlines()\n",
    "Lines_R2 = file_R2.readlines()\n",
    "\n",
    "new_Lines_R1 = []\n",
    "new_Lines_R2 = []\n",
    "\n",
    "complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "\n",
    "# Pick up sequences from FASTQ file and preserve the first 151 nt in case some oligo lengths were beyond 151 nt.\n",
    "for i in range (0, int(len(Lines_R1))):\n",
    "    if i%4 == 1:\n",
    "        new_Lines_R1.append(Lines_R1[i][0:151].strip('\\n'))\n",
    "\n",
    "for i in range (0, int(len(Lines_R2))):\n",
    "    if i%4 == 1:\n",
    "        reverse_complement_R2 = \"\".join(complement.get(base, base) for base in reversed(Lines_R2[i][0:151].strip('\\n')))\n",
    "        new_Lines_R2.append(reverse_complement_R2)\n",
    "        \n",
    "print('Total number of reads: ' + str(len(new_Lines_R1)))      \n",
    "\n",
    "new_Lines_Combo = []\n",
    "\n",
    "# Define a list to store Read Number of those reads which passed the following conditions:\n",
    "read_index_1 = []\n",
    "\n",
    "for i in range (0, int(len(new_Lines_R1))):\n",
    "    # Because the length of Reference Strand is less than 151 nt, we only need to find if 142 nt of the forward read and reverse read overlap\n",
    "    if (new_Lines_R1[i][0:142] == new_Lines_R2[i][9:151]):  # This judgement is to ensure that the paired-end read generates consistent results, therefore enhance the probability that this read is valid.\n",
    "        new_Lines_Combo.append(new_Lines_R1[i][0:142])\n",
    "        read_index_1.append(i)\n",
    "    \n",
    "Lines = new_Lines_Combo\n",
    "print('Total number of reads after processing: ' + str(len(Lines)))\n",
    "\n",
    "# Get the number of valid reads\n",
    "res_read_index_1 = [*set(read_index_1)]\n",
    "        \n",
    "primerLength = 21\n",
    "segmentLimit = 9\n",
    "referenceLimit = 16\n",
    "referenceStrands = [[None]*segmentLimit for _ in range(referenceLimit)]\n",
    "candi_referenceStrands = [ [ [] for i in range(segmentLimit) ] for i in range(referenceLimit) ]\n",
    "\n",
    "read_index_2 = []\n",
    "read_index_3 = []\n",
    "\n",
    "read_number = -1\n",
    "\n",
    "for reference in Lines: \n",
    "    read_number += 1\n",
    "    arrangedReference = []\n",
    "    reference = reference.strip('\\n')   # In '.txt' files, there may be '\\n' symbols meaning the start of a new line. Those symbols needs to be eliminated. \n",
    "    # Exclude forward and reverse primers from Reference Strands\n",
    "    reference = reference[primerLength:len(reference)-primerLength]\n",
    "\n",
    "    RS = reference[-12:]      # Extract RS Sequence (base-3)\n",
    "\n",
    "    reference = reference[0:(len(reference)-12)]\n",
    "    compare_candidate = reference[8:]\n",
    "    \n",
    "    for m in range (0, referenceLimit):\n",
    "        for n in range (0, segmentLimit):\n",
    "            if compare_candidate == array_reference[m][n]:\n",
    "                read_index_2.append(read_index_1[read_number])   # Store Read Number with successful decoding attempt before RS correction\n",
    "\n",
    "    # Convert RS Sequence to ternary number, then to binary number\n",
    "    RS_segment_binary_total = []\n",
    "    for i in range (0, int(len(RS)/6)):\n",
    "        RS_segment = RS[i*6:(i+1)*6]\n",
    "        RS_segment_base3 = oligoToBase3(RS_segment)\n",
    "        if (RS_segment_base3 == -1):\n",
    "            break\n",
    "        RS_segment_decimal = ternaryToDecimal(RS_segment_base3)\n",
    "        RS_segment_binary = bin(RS_segment_decimal)\n",
    "        RS_segment_binary = RS_segment_binary[2:]\n",
    "        RS_segment_binary = '0'*(8-len(RS_segment_binary)) + RS_segment_binary\n",
    "        RS_segment_binary = str(RS_segment_binary)\n",
    "        RS_segment_binary_total.append(RS_segment_binary)\n",
    "    \n",
    "\n",
    "    RS_segment_binary_total = ''.join(RS_segment_binary_total)\n",
    "    if (len(RS_segment_binary_total) != referenceLimit):\n",
    "        continue\n",
    "        \n",
    "    binaryConverted = converter(reference)   # Convert Reference Sequence to binary number \n",
    "    binaryConverted = binaryConverted + RS_segment_binary_total   # Combine the reference and RS code in binary form\n",
    "\n",
    "    binaryList = [int(binaryConverted[i:i + 8], 2) for i in range(0, len(binaryConverted), 8)]\n",
    "    bytesList = bytes(binaryList)\n",
    "    try:\n",
    "        RSDecoded = rsc.decode(bytesList)[0]   # RS correction\n",
    "    except:\n",
    "        continue\n",
    "    \n",
    "    bytes_as_bits = ''.join(format(byte, '08b') for byte in RSDecoded)\n",
    "    baseDeconverted = deconverter(bytes_as_bits)   # Convert corrected bytes back to DNA sequences \n",
    "\n",
    "    referenceSequence = baseDeconverted[0:4]    # Get Reference Number\n",
    "    segmentSequence = baseDeconverted[4:8]      # Get Segment Number\n",
    "    referenceBase3 = oligoToBase3(referenceSequence)\n",
    "    if (referenceBase3 == -1):\n",
    "        continue\n",
    "    segmentBase3 = oligoToBase3(segmentSequence)\n",
    "    if(segmentBase3 == -1):\n",
    "        continue\n",
    "\n",
    "    referenceBase10 = ternaryToDecimal(referenceBase3)\n",
    "    segmentBase10 = ternaryToDecimal(segmentBase3)\n",
    "    \n",
    "    baseDeconverted = baseDeconverted[8:]   # Get Reference Sequence\n",
    "\n",
    "    if (len(baseDeconverted) != 80):\n",
    "        continue\n",
    "        \n",
    "    for m in range (0, referenceLimit):\n",
    "        for n in range (0, segmentLimit):\n",
    "            if baseDeconverted == array_reference[m][n]:\n",
    "                read_index_3.append(read_index_1[read_number])   # Store Read Number with successful decoding attempt before RS correction\n",
    "    try:\n",
    "        candi_referenceStrands[referenceBase10][segmentBase10].append(baseDeconverted) \n",
    "        \n",
    "    except:\n",
    "        continue\n",
    "        \n",
    "for m in range (0, referenceLimit):\n",
    "    for n in range (0, segmentLimit):\n",
    "        if (referenceStrands[m][n] == None):\n",
    "            try:\n",
    "                referenceStrands[m][n] = most_frequent(candi_referenceStrands[m][n])  # Only keep the sequence appearing most frequently\n",
    "            except:\n",
    "                pass\n",
    "        \n",
    "# Due to repeated \"CGTA\" unkts in the Reference Strands, some sequences may not be decodable even if we enhance the sequencing coverage. In this case, a remedial mechanism may need to be introduced.        \n",
    "# To determine the unique repeat sequence in one Reference Segment (first 4-letter decodeable unit)\n",
    "\n",
    "\n",
    "# Get the number of successfully decoded reads before and after RS correction\n",
    "res_read_index_2 = [*set(read_index_2)]\n",
    "res_read_index_3 = [*set(read_index_3)]\n",
    "print('Before Remedy: ') \n",
    "print('Total number of valid reads: ' + str(len(res_read_index_1)))\n",
    "print('Exact matched sequences before RS correction: ' + str(len(res_read_index_2)))\n",
    "print('Matched sequences after RS correction: ' + str(len(res_read_index_3)))\n",
    "\n",
    "# Get the index list of non-decodable reads\n",
    "whole_read = list(range(len(new_Lines_R1)))\n",
    "not_decodable_index = [element for element in whole_read if element not in res_read_index_2]\n",
    "\n",
    "# Define correctable remedy reads\n",
    "remedy_reads = []\n",
    "# Define a list to store Read Number of those reads which passed the following conditions\n",
    "read_index_4 = []   \n",
    "\n",
    "candi_referenceStrands = [ [ [] for i in range(segmentLimit) ] for i in range(referenceLimit) ]\n",
    "    \n",
    "for i in range(0, len(not_decodable_index)):\n",
    "    candidate = new_Lines_R1[not_decodable_index[i]]  # Pick up candidate reads according to the read index\n",
    "    for j in range (0, len(candidate)):\n",
    "        if candidate[0:21] == 'TCAACTGGTGATTCGTGCAAC' and candidate[j:j+21] == 'ACGGTAGCTTCCTGTATGCCT':  # Recognition of reverse primer sequence\n",
    "            remedy_reads.append(candidate[0:j+21])  \n",
    "            read_index_4.append(not_decodable_index[i])\n",
    "            \n",
    "# Supplement remedy reads to a length of 80 nt            \n",
    "remedy_reads_corrected = []\n",
    "for reference in remedy_reads:\n",
    "    \n",
    "    reference = reference[21:len(reference)-21]  # Exclude PCR forward and reverse primers\n",
    "    RS = reference[-12:]\n",
    "    referenceSequence = reference[0:4]\n",
    "    segmentSequence = reference[4:8]\n",
    "    reference = reference[8:len(reference)-12]   # Get payload of Reference\n",
    "    \n",
    "    # Base supplementing rule\n",
    "    if (len(reference) == 0):\n",
    "        reference = 20 * 'CGTA'\n",
    "        \n",
    "    if len(reference)%4 == 0:\n",
    "        reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        \n",
    "    elif len(reference)%4 == 1:\n",
    "        if (reference[-1:]) == \"C\":\n",
    "            reference = reference + 'GTA'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-1:]) == \"G\":\n",
    "            reference = reference + 'TAC'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-1:]) == \"T\":\n",
    "            reference = reference + 'ACG'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-1:]) == \"A\":\n",
    "            reference = reference + 'CGT'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "            \n",
    "    elif len(reference)%4 == 2:\n",
    "        if (reference[-2:]) == \"CG\":\n",
    "            reference = reference + 'TA'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-2:]) == \"GT\":\n",
    "            reference = reference + 'AC'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-2:]) == \"TA\":\n",
    "            reference = reference + 'CG'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-2:]) == \"AC\":\n",
    "            reference = reference + 'GT'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        \n",
    "    elif len(reference)%4 == 3:\n",
    "        if (reference[-3:]) == \"CGT\":\n",
    "            reference = reference + 'A'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-3:]) == \"GTA\":\n",
    "            reference = reference + 'C'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-3:]) == \"TAC\":\n",
    "            reference = reference + 'G'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "        elif (reference[-3:]) == \"ACG\":\n",
    "            reference = reference + 'T'\n",
    "            reference = reference + reference[-4:]*(int((80-len(reference))/4))\n",
    "    \n",
    "    if (len(reference) == 80):\n",
    "        reference = referenceSequence + segmentSequence + reference + RS\n",
    "        remedy_reads_corrected.append(reference)\n",
    "        \n",
    "# Go back to normal processing \n",
    "read_index_5 = []\n",
    "read_index_6 = []\n",
    "\n",
    "read_number = -1\n",
    "\n",
    "for reference in remedy_reads_corrected:\n",
    "    read_number += 1\n",
    "    RS = reference[-12:]      # Extract RS Sequence (base-3)\n",
    "\n",
    "    reference = reference[0:(len(reference)-12)]\n",
    "    \n",
    "    compare_candidate = reference[8:]\n",
    "    \n",
    "    for m in range (0, referenceLimit):\n",
    "        for n in range (0, segmentLimit):\n",
    "            if compare_candidate == array_reference[m][n]:\n",
    "                read_index_5.append(read_index_4[read_number])   # Store Read Number with successful decoding attempt before RS correction\n",
    "\n",
    "    # Convert RS Sequence to ternary number, then to binary number\n",
    "    RS_segment_binary_total = []\n",
    "    for i in range (0, int(len(RS)/6)):\n",
    "        RS_segment = RS[i*6:(i+1)*6]\n",
    "        RS_segment_base3 = oligoToBase3(RS_segment)\n",
    "        if (RS_segment_base3 == -1):\n",
    "            break\n",
    "        RS_segment_decimal = ternaryToDecimal(RS_segment_base3)\n",
    "        RS_segment_binary = bin(RS_segment_decimal)\n",
    "        RS_segment_binary = RS_segment_binary[2:]\n",
    "        RS_segment_binary = '0'*(8-len(RS_segment_binary)) + RS_segment_binary\n",
    "        RS_segment_binary = str(RS_segment_binary)\n",
    "        RS_segment_binary_total.append(RS_segment_binary)\n",
    "    \n",
    "\n",
    "    RS_segment_binary_total = ''.join(RS_segment_binary_total)\n",
    "    if (len(RS_segment_binary_total) != referenceLimit):\n",
    "        continue\n",
    "        \n",
    "    binaryConverted = converter(reference)   # Convert Reference Sequence to binary number \n",
    "    binaryConverted = binaryConverted + RS_segment_binary_total   # Combine the reference and RS code in binary form\n",
    "\n",
    "    binaryList = [int(binaryConverted[i:i + 8], 2) for i in range(0, len(binaryConverted), 8)]\n",
    "    bytesList = bytes(binaryList)\n",
    "    try:\n",
    "        RSDecoded = rsc.decode(bytesList)[0]   # RS correction\n",
    "    except:\n",
    "        continue\n",
    "    \n",
    "    bytes_as_bits = ''.join(format(byte, '08b') for byte in RSDecoded)\n",
    "    baseDeconverted = deconverter(bytes_as_bits)   # Convert corrected bytes back to DNA sequences \n",
    "\n",
    "    referenceSequence = baseDeconverted[0:4]    # Get Reference Number\n",
    "    segmentSequence = baseDeconverted[4:8]      # Get Segment Number\n",
    "    referenceBase3 = oligoToBase3(referenceSequence)\n",
    "    if (referenceBase3 == -1):\n",
    "        continue\n",
    "    segmentBase3 = oligoToBase3(segmentSequence)\n",
    "    if (segmentBase3 == -1):\n",
    "        continue\n",
    "\n",
    "    referenceBase10 = ternaryToDecimal(referenceBase3)\n",
    "    segmentBase10 = ternaryToDecimal(segmentBase3)\n",
    "    \n",
    "    baseDeconverted = baseDeconverted[8:]   # Get Reference Sequence\n",
    "\n",
    "    if (len(baseDeconverted) != 80):\n",
    "        continue\n",
    "        \n",
    "    for m in range (0, referenceLimit):\n",
    "        for n in range (0, segmentLimit):\n",
    "            if baseDeconverted == array_reference[m][n]:\n",
    "                read_index_6.append(read_index_4[read_number])   # Store Read Number with successful decoding attempt before RS correction\n",
    "    try:\n",
    "        if (referenceStrands[referenceBase10][segmentBase10] == None):\n",
    "            candi_referenceStrands[referenceBase10][segmentBase10].append(baseDeconverted) # Store the RS-corrected sequence in candidate sequence list\n",
    "    \n",
    "    except:\n",
    "        continue\n",
    "\n",
    "\n",
    "for i in range (0, referenceLimit):\n",
    "    for j in range (0, segmentLimit):\n",
    "        if (referenceStrands[i][j] == None):\n",
    "            referenceStrands[i][j] = most_frequent(candi_referenceStrands[i][j])  # Only keep the sequence appearing most frequently\n",
    "            \n",
    "  \n",
    "for i in range (0, len(referenceStrands)):\n",
    "    try:\n",
    "        referenceStrands[i] = ''.join(referenceStrands[i])    # Put together reference strands\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "print(referenceStrands)\n",
    "\n",
    "%store referenceStrands      \n",
    "%store -r trimmed_whole_array\n",
    "# Compare to standard \n",
    "for i in range (0, len(trimmed_whole_array)):\n",
    "    if (trimmed_whole_array[i] == referenceStrands[i]):\n",
    "        print(\"Reference #\" + str(i) + ': Yes!')\n",
    "    else:\n",
    "        print(\"Reference #\" + str(i) + ': No!')   \n",
    "    \n",
    "# Get the number of successfully decoded reads before and after RS correction\n",
    "res_read_index_5 = [*set(read_index_5)]\n",
    "res_read_index_6 = [*set(read_index_6)]\n",
    "print('Remedy Process: ')\n",
    "print('Total number of valid reads: ' + str(len(read_index_4)))\n",
    "print('Exact matched sequences before RS correction: ' + str(len(res_read_index_5)))\n",
    "print('Matched sequences after RS correction: ' + str(len(res_read_index_6)))       \n",
    "\n",
    "print('After Remedy: ')    \n",
    "print('Total number of valid reads: ' + str(len(res_read_index_1) + len(read_index_4)))\n",
    "print('Exact matched sequences before RS correction: ' + str(len(res_read_index_2) + len(res_read_index_5)))\n",
    "print('Matched sequences after RS correction: ' + str(len(res_read_index_3) + len(res_read_index_6)))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
